File: question/neetcode/binary_tree/binary-tree-level-order-traversal.py
"""
        1
    2       3
4       5

ans = []

dfs(node, level)
dfs(root, 0)

ans = [[1], [2,3]]
1 -> call 2,3
"""


from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        ans = []

        def dfs(node: Optional[TreeNode], level: int, ans: List[List[int]]):
            if not node:
                return
            
            while len(ans) <= level:
                ans.append([])

            ans[level].append(node.val)
            dfs(node.left, 1+level, ans)
            dfs(node.right, 1+level, ans)

        dfs(root, 0, ans)

        return ans

File: question/neetcode/binary_tree/binary-tree-level-order-traversal.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode*node, vector<vector<int>>& ans, int d){
        if(not node) return;

        if(ans.size()==d){
            ans.push_back({});
        } 
        ans[d].push_back(node->val);
        dfs(node->left,ans,d+1);
        dfs(node->right,ans,d+1);
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        dfs(root, ans, 0);

        return ans;
    }
};